In [1]:
# The first step of using numpy is to tell python to import it
import numpy as np

### 2.1 NumPy Arrays

**NumPy Array**
* An array is a data structure that stores values of same data type.
* While python lists can contain values corresponding to different data types, arrays in python can only contain values corresponding to the same data type. 
* However python lists fail to deliver the performance required while computing large sets of numerical data. To address this issue we use NumPy arrays.
* We can create NumPy arrays by converting a list to an array.


In [ ]:
# defining a list of different car companies or string elements
arr_str = ['Mercedes', 'BMW', 'Audi', 'Ferrari', 'Tesla']

# defining a list of number of cylinders in car or numerical elements
arr_num = [5, 4, 6, 7, 6.0,True,"saif"]

In [15]:
# connverting the list arr_str to a NumPy array
np_arr_str = np.array(arr_str)


# connverting the list arr_num to a NumPy array
np_arr_num = np.array(arr_num)

# checking the output
print('Numpy Array (arr_str): ',np_arr_str)
print('Numpy Array (arr_num): ',np_arr_num)

Numpy Array (arr_str):  ['Mercedes' 'BMW' 'Audi' 'Ferrari' 'Tesla']
Numpy Array (arr_num):  ['5' '4' '6' '7' '6.0' 'True' 'saif']


The resuts look similar to a list but arr_str and arr_num have been converted to NumPy arrays. Let's check the data type to confirm this.

In [16]:
# printing the data type of lists
print('Data type of arr_str: ',type(arr_str))
print('Data type of arr_num: ',type(arr_num))

# printing the data type after conversion of lists to array
print('Data type of np_arr_str: ',type(np_arr_str))
print('Data type of np_arr_num: ',type(np_arr_num))

Data type of arr_str:  <class 'list'>
Data type of arr_num:  <class 'list'>
Data type of np_arr_str:  <class 'numpy.ndarray'>
Data type of np_arr_num:  <class 'numpy.ndarray'>


* The above output confirms that both the lists were successfully converted to arrays

**NumPy Matrix**

* A matrix is a two-dimensional data structure where elements are arranged into rows and columns.
* A matrix can be created by using list of lists

In [6]:
lists=[[1,2,1],[4,5,9],[1,8,9]]
# let's say we have information of different number of cylinders in a car and we want to display them in a matrix format
matrix = np.array(lists)
print(matrix)

[[1 2 1]
 [4 5 9]
 [1 8 9]]


In [8]:
print('Data type of matrix: ',type(matrix))

Data type of matrix:  <class 'numpy.ndarray'>


* We see that all the NumPy objects have data type as ndarray

### 2.2 NumPy Functions

**There are different ways to create NumPy arrays using the functions available in NumPy library**

**Using np.arange() function**
* The np.arange() function returns an array with evenly spaced elements as per the interval. The interval mentioned is half-opened i.e. start is included but stop is excluded.
* It has the following paramaters:
  * start : start of interval range. By default start = 0
  * stop  : end of interval range
  * step  : step size of interval. By default step size = 1

In [7]:
arr2  = np.arange(start = 0, stop = 10) # 10 will be excluded from the output
print(arr2)

# or

arr2  = np.arange(0,10) 
print(arr2)

[0 1 2 3 4 5 6 7 8 9]
[0 1 2 3 4 5 6 7 8 9]


In [8]:
# adding a step size of 5 to create an array
arr3  = np.arange(start = 0, stop = 20, step = 5)
arr3

array([ 0,  5, 10, 15])

**Using np.linspace() function**
* The np.linspace() function returns numbers which are evenly distributed with respect to interval. Here the start and stop both are included.            
*It has the following parameters:              
 * start: start of interval range. By default start = 0
 * stop: end of interval range
 * num : No. of samples to generate. By default num = 50

In [21]:
matrix2 = np.linspace(0,5) # by default 50 evenly spaced values will be generated between 0 and 5
matrix2

array([0.        , 0.10204082, 0.20408163, 0.30612245, 0.40816327,
       0.51020408, 0.6122449 , 0.71428571, 0.81632653, 0.91836735,
       1.02040816, 1.12244898, 1.2244898 , 1.32653061, 1.42857143,
       1.53061224, 1.63265306, 1.73469388, 1.83673469, 1.93877551,
       2.04081633, 2.14285714, 2.24489796, 2.34693878, 2.44897959,
       2.55102041, 2.65306122, 2.75510204, 2.85714286, 2.95918367,
       3.06122449, 3.16326531, 3.26530612, 3.36734694, 3.46938776,
       3.57142857, 3.67346939, 3.7755102 , 3.87755102, 3.97959184,
       4.08163265, 4.18367347, 4.28571429, 4.3877551 , 4.48979592,
       4.59183673, 4.69387755, 4.79591837, 4.89795918, 5.        ])

**How are these values getting generated?**

The step size or the difference between each element will be decided by the following formula:

**(stop - start) / (total elements - 1)**

So, in this case:
(5 - 0) / 49 = 0.10204082

The first value will be 0.10204082, the second value will be 0.10204082 + 0.10204082, the third value will be 0.10204082 + 0.10204082 +0.10204082, and so on.

In [13]:
# generating 10 evenly spaced values between 10 and 20
matrix3 = np.linspace(10,20,9)
matrix3

array([10.  , 11.25, 12.5 , 13.75, 15.  , 16.25, 17.5 , 18.75, 20.  ])

**Similarly we can create matrices using the functions available in NumPy library**

**Using np.zeros()**
 
* The np.zeros() is a function for creating a matrix and performing matrix operations in NumPy. 
* It returns a matrix filled with zeros of the given shape. 
* It has the following parameters:    
  * shape : Number of rows and columns in the output matrix.
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [23]:
matrix4 = np.zeros([2,5])
matrix4

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

**Using np.ones()**

* The np.ones() is another function for creating a matrix and performing matrix operations in NumPy. 
* It returns a matrix of given shape and type, filled with ones.
* It has the following parameters:  
  * shape : Number of rows and columns in the output matrix.
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [25]:
matrix5 = np.ones([3,5])
matrix5

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

**Using np.eye()**
* The np.eye() is a function for creating a matrix and performing matrix operations in NumPy. 
* It returns a matrix with ones on the diagonal and zeros elsewhere. 
* It has the following parameters:
  * n: Number of rows and columns in the output matrix 
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [26]:
matrix6 = np.eye(5)
matrix6

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

**We can also convert a one dimension array to a matrix. This can be done by using the np.reshape() function.**

* The shape of an array basically tells the number of elements and dimensions of the array. Reshaping a Numpy array simply means changing the shape of the given array. 
* By reshaping an array we can add or remove dimensions or change number of elements in each dimension. 
* In order to reshape a NumPy array, we use the reshape method with the given array. 
* **Syntax:** array.reshape(shape) 
  * shape: a tuple given as input, the values in tuple will be the new shape of the array.

In [28]:
# defining an array with values 0 to 9
arr4 = np.arange(0,10) 
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [34]:
# reshaping the array arr4 to a 2 x 5 matrix
arr4_reshaped = arr4.reshape((2,5))
arr4_reshaped

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [18]:
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [19]:
# reshaping the array arr4 to a 2 x 6 matrix
arr4.reshape((2,6))

ValueError: ignored

* This did not work because we have 10 elements which we are trying to fit in a 2 X 6 shape which will require 12 elements.

**NumPy can also perform a large number of different mathematical operations and it provides different functions to do so.**

NumPy provides:
1. Trigonometric functions 
2. Exponents and Logarithmic functions
3. Functions for arithmetic operations between arrays and matrices

**Trigonometric functions**

In [20]:
print('Sine Function:',np.sin(4))
print('Cosine Function:',np.cos(4))
print('Tan Function',np.tan(4))

Sine Function: -0.7568024953079282
Cosine Function: -0.6536436208636119
Tan Function 1.1578212823495775


**Exponents and Logarithmic functions**

* Exponents

In [21]:
np.exp(2)

7.38905609893065

In [22]:
arr5 = np.array([2,4,6])
np.exp(arr5)

array([  7.3890561 ,  54.59815003, 403.42879349])

* Logarithms

In [23]:
# by default NumPy takes the base of log as e
np.log(2)

0.6931471805599453

In [24]:
np.log(arr5)

array([0.69314718, 1.38629436, 1.79175947])

In [25]:
## log with base 10
np.log10(8) 

0.9030899869919435

**Arithmetic Operations on arrays**

In [15]:
# arithmetic on lists

l1 = [1,2,3]
l2 = [4,5,6]
print(l1+l2)
# this does not behave as you would expect!


[1, 2, 3, 4, 5, 6]


In [37]:
# we can +-*/ arrays together

# defining two arrays
arr7 = np.arange(1,6)
print('arr7:', arr7)

arr8 = np.arange(3,8)
print('arr8:', arr8)

arr7: [1 2 3 4 5]
arr8: [3 4 5 6 7]


In [38]:
print('Addition: ',arr7+arr8)
print('Subtraction: ',arr8-arr7)
print('Multiplication:' , arr7*arr8)
print('Division:', arr7/arr8)
print('Inverse:', 1/arr7)
print('Powers:', arr7**arr8) # in python, powers are achieved using **.


Addition:  [ 4  6  8 10 12]
Subtraction:  [2 2 2 2 2]
Multiplication: [ 3  8 15 24 35]
Division: [0.33333333 0.5        0.6        0.66666667 0.71428571]
Inverse: [1.         0.5        0.33333333 0.25       0.2       ]
Powers: [    1    16   243  4096 78125]


**Operations on Matrices**

In [20]:
matrix7 = np.arange(1,10) # signle dimensional array
matrix7=matrix7.reshape(3,3) # two dimensional array
print(matrix7)

matrix8 = np.eye(3)
print(matrix8)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [30]:
print('Addition: \n', matrix7+matrix8)
print('Subtraction: \n ', matrix7-matrix8)
print('Multiplication: \n', matrix7*matrix8)
print('Division: \n', matrix7/matrix8)

Addition: 
 [[ 2.  2.  3.]
 [ 4.  6.  6.]
 [ 7.  8. 10.]]
Subtraction: 
  [[0. 2. 3.]
 [4. 4. 6.]
 [7. 8. 8.]]
Multiplication: 
 [[1. 0. 0.]
 [0. 5. 0.]
 [0. 0. 9.]]
Division: 
 [[ 1. inf inf]
 [inf  5. inf]
 [inf inf  9.]]


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: RuntimeWarning: divide by zero encountered in true_divide
  after removing the cwd from sys.path.


* RuntimeWarning: Errors which occur during program execution(run-time) after successful compilation are called run-time errors. 
* One of the most common run-time error is division by zero also known as Division error. 
* Due to division by zero error, we are getting inf (infinity) values because 1/0 is not a defined operation.

**Linear algebra matrix multiplication**

In [31]:
matrix9 = np.arange(1,10).reshape(3,3)
print('First Matrix: \n',matrix9)

matrix10 = np.arange(11,20).reshape(3,3)
print('Second Matrix: \n',matrix10)
print('')
# taking linear algebra matrix multiplication (some may have heard this called the dot product)
print('Multiplication: \n', matrix9 @ matrix10)

First Matrix: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Second Matrix: 
 [[11 12 13]
 [14 15 16]
 [17 18 19]]

Multiplication: 
 [[ 90  96 102]
 [216 231 246]
 [342 366 390]]


**Transpose of a matrix**

In [32]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [33]:
# taking transpose of matrix
np.transpose(matrix9)

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [34]:
# another way of taking a transpose
matrix9.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

**Function to find minimum and maximum values**

In [35]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [36]:
print('Minimum value: ',np.min(matrix9))

Minimum value:  1


In [37]:
print('Maximum value: ',np.max(matrix9))

Maximum value:  9


**Function to generate random samples**

**Using np.random.rand function**

* The np.random.rand returns a random NumPy array whose element(s) are drawn randomly from the normal distribution over [0,1). (including 0 but excluding 1). 
* **Syntax** - np.random.rand(d0,d1)
  * d0,d1 – It represents the dimension of the required array given as int, where d1 is optional. 

In [39]:
# Generating random values in an array
rand_mat = np.random.rand(5)
print(rand_mat)

[0.34838911 0.41842297 0.42809383 0.58933788 0.11477283]


In [25]:
# * Generating random values in a matrix
rand_mat = np.random.rand(5,5) # uniform random variable
print(rand_mat)

[[0.63388585 0.82033537 0.53199424 0.30979862 0.5615088 ]
 [0.35242894 0.77485452 0.13708448 0.37573045 0.09893122]
 [0.75582359 0.65020809 0.98678176 0.59698634 0.42286669]
 [0.74596416 0.70889014 0.70870177 0.42497425 0.06260179]
 [0.10607269 0.62160539 0.25445013 0.20078087 0.82392644]]


**Using np.random.randn function**

* The np.random.randn returns a random numpy array whose sample(s) are drawn randomly from the standard normal distribution (Mean as 0 and standard deviation as 1)

* **Syntax** - np.random.randn(d0,d1)
  * d0,d1 – It represents the dimension of the output, where d1 is optional.

In [42]:
# Generating random values in an array
rand_mat2 = np.random.randn(40) 
print(rand_mat2)

np.mean(rand_mat2)

[ 0.27264653  0.3062506  -0.46512396  1.44243326 -0.74023543 -1.89683498
 -1.43931479  2.22475696  0.6648984  -1.42807979  0.93957976  0.32433547
  0.05621024  0.82665009 -0.56050265  0.13323254 -0.81512074  0.44860705
  1.3384705  -0.97233228 -0.05169043  0.59673441  1.66513503  0.68259385
 -0.50300972  2.55897028  1.97483011  0.06384297 -0.747679    1.03940902
 -0.52277389 -1.32299113  0.05481717 -0.2486384   0.72663359 -0.35848656
  1.05412558  0.72517254  0.15621022  0.19934706]


0.21007698726049182

In [43]:
# Generating random values in a matrix
rand_mat2 = np.random.randn(5,5) 
print(rand_mat2)

[[ 1.69744907 -2.56880637  0.09779989 -0.38044146 -0.09819401]
 [-1.40811811  0.87066695  0.9600046  -0.23449706  1.49096701]
 [-0.04018511 -0.16657884  0.09398931  1.2705447   0.53368585]
 [-0.57503518  0.72198025 -0.94105769 -1.31375521 -0.19174442]
 [ 1.80220964 -1.31644501 -0.84424066  0.13730161 -0.79708007]]


In [33]:
# Let's check the mean and standard deviation of rand_mat2
print('Mean:',np.mean(rand_mat2))
print('Standard Deviation:',np.std(rand_mat2))

Mean: 0.07533262288448284
Standard Deviation: 1.0859461074767993


*  We observe that the mean is very close to 0 and standard deviation is very close to 1.

**Using np.random.randint function**

* The np.random.randint returns a random numpy array whose element(s) are drawn randomly from low (inclusive) to the high (exclusive) range. 

* **Syntax** - np.random.randint(low, high, size) 

  * low – It represents the lowest inclusive bound of the distribution from where the sample can be drawn.
  * high – It represents the upper exclusive bound of the distribution from where the sample can be drawn. 
  * size – It represents the shape of the output. 

In [36]:
# Generating random values in an array
rand_mat3 = np.random.randint(1,5,10)
print(rand_mat3)

[4 4 2 4 2 1 4 2 4 2]


In [37]:
# Generating random values in a matrix
rand_mat3 = np.random.randint(1,40,[5,5])
print(rand_mat3)

[[19 28 35 12 16]
 [21 33 37 19 20]
 [20 22 25 16 27]
 [21 20  6 39 38]
 [32 36  6 37 39]]


### 2.3 Accessing the entries of a Numpy Array

In [40]:
# let's generate an array with 10 random values
rand_arr = np.random.randn(10)
print(rand_arr)

[ 0.67448984  1.1657466  -0.4655218  -1.17819306  1.37430978 -1.04228452
 -0.46058435  0.66505119 -0.40385783  1.56610372]


* Accessing one element from an array

In [41]:
# accessing the 6 th entry of rand_arr
print(rand_arr[6])

-0.4605843523089329


* Accessing multiple elements from an array

In [42]:
# we can access multiple entries at once using
print(rand_arr[4:9])

[ 1.37430978 -1.04228452 -0.46058435  0.66505119 -0.40385783]


In [47]:
# we can also access multiple non-consecutive entries using np.arange
print('Index of values to access: ',np.arange(3,10,3))
print(rand_arr[np.arange(3,10,3)])

Index of values to access:  [3 6 9]
[-1.17819306 -0.46058435  1.56610372]


**Accessing arrays using logical operations**

In [49]:
print(rand_arr)

[-2.04194609  0.45031706 -0.02019619  1.40439998  0.27655997  0.78027138
 -0.95580464  0.05992533  0.16359666  0.19479355]


In [50]:
rand_arr>0

array([False,  True, False,  True,  True,  True, False,  True,  True,
        True])

In [51]:
# accessing all the values of rand_arr which are greater than 0
print('Values greater than 0: ',rand_arr[rand_arr>0])

# accessing all the values of rand_arr which are less than 0
print('Values less than 0: ',rand_arr[rand_arr<0])

Values greater than 0:  [0.45031706 1.40439998 0.27655997 0.78027138 0.05992533 0.16359666
 0.19479355]
Values less than 0:  [-2.04194609 -0.02019619 -0.95580464]


**Accessing the entries of a Matrix**

In [79]:
# let's generate an array with 10 random values
rand_mat = np.random.randn(5,5)
print(rand_mat)

[[-0.82143833 -0.06320327  2.02286993 -0.29794316  0.80229657]
 [-0.13884546  1.15734187  0.28569263 -0.42192284 -0.36478021]
 [-0.07835671  2.11000357 -0.9506834   0.42590556 -0.45060251]
 [-0.16245548 -0.46694343 -0.15550438  0.31551549 -1.72685098]
 [ 0.70500886 -2.17388943 -0.13136581 -0.73299754  0.42873511]]


In [50]:
# acessing the second row of the rand_mat
rand_mat[1]

array([ 0.05921141, -0.13788599,  0.3571995 ,  0.91453566, -1.70747721])

In [51]:
# acessing third element of the second row
print(rand_mat[1][2])

#or 

print(rand_mat[2,4])

0.35719949622538927
1.9993322557186703


In [52]:
# accessing first two rows with second and third column 
print(rand_mat[0:2,1:3])

[[-0.45791714  0.78257068]
 [-0.13788599  0.3571995 ]]


**Accessing matrices using logical operations**

In [53]:
print(rand_mat)

[[ 1.85490064 -0.45791714  0.78257068 -1.17791506 -0.98282917]
 [ 0.05921141 -0.13788599  0.3571995   0.91453566 -1.70747721]
 [-0.5522617   0.0689841  -0.50539164 -0.72530254  1.99933226]
 [ 1.50127045 -1.19402222  0.33829004 -0.34925233  0.41117229]
 [-0.21511244  0.38073838  0.24985351 -0.66315327 -0.85488619]]


In [54]:
# accessing all the values of rand_mat which are greater than 0
print('Values greater than 0: \n ',rand_mat[rand_mat>0])

# accessing all the values of rand_mat which are less than 0
print('Values less than 0: \n',rand_mat[rand_mat<0])

Values greater than 0: 
  [1.85490064 0.78257068 0.05921141 0.3571995  0.91453566 0.0689841
 1.99933226 1.50127045 0.33829004 0.41117229 0.38073838 0.24985351]
Values less than 0: 
 [-0.45791714 -1.17791506 -0.98282917 -0.13788599 -1.70747721 -0.5522617
 -0.50539164 -0.72530254 -1.19402222 -0.34925233 -0.21511244 -0.66315327
 -0.85488619]


**Modifying the entries of an Array**

In [55]:
print(rand_arr)


[ 0.67448984  1.1657466  -0.4655218  -1.17819306  1.37430978 -1.04228452
 -0.46058435  0.66505119 -0.40385783  1.56610372]


In [56]:
# let's change some values in an array!
# changing the values of index value 3 and index value 4 to 5
rand_arr[3:5] = 5
print(rand_arr)

[ 0.67448984  1.1657466  -0.4655218   5.          5.         -1.04228452
 -0.46058435  0.66505119 -0.40385783  1.56610372]


In [58]:
# changing the values of index value 0 and index value 1 to 2 and 3 respectively
rand_arr[0:3] = [2,3,4]
print(rand_arr)

[ 2.          3.          4.          5.          5.         -1.04228452
 -0.46058435  0.66505119 -0.40385783  1.56610372]


In [59]:
# modify entries using logical references
rand_arr[rand_arr>0] = 65
rand_arr

array([65.        , 65.        , 65.        , 65.        , 65.        ,
       -1.04228452, -0.46058435, 65.        , -0.40385783, 65.        ])

**Modifying the entries of a Matrix**

In [60]:
print(rand_mat3)

[[19 28 35 12 16]
 [21 33 37 19 20]
 [20 22 25 16 27]
 [21 20  6 39 38]
 [32 36  6 37 39]]


In [61]:
# changing the values of the 4th and 5th element of the second and third rows of the matrix to 0
print('Matrix before modification: \n',rand_mat3)
rand_mat3[1:3,3:5] = 0
print('Matrix after modification: \n',rand_mat3)

Matrix before modification: 
 [[19 28 35 12 16]
 [21 33 37 19 20]
 [20 22 25 16 27]
 [21 20  6 39 38]
 [32 36  6 37 39]]
Matrix after modification: 
 [[19 28 35 12 16]
 [21 33 37  0  0]
 [20 22 25  0  0]
 [21 20  6 39 38]
 [32 36  6 37 39]]


In [80]:
# extracting the first 2 rows and first 3 columns from the matrix
sub_mat = rand_mat[0:2,0:3]
print(sub_mat)

[[-0.82143833 -0.06320327  2.02286993]
 [-0.13884546  1.15734187  0.28569263]]


In [81]:
# changing all the values of the extracted matrix to 3
sub_mat[:] = 3
print(sub_mat)

[[3. 3. 3.]
 [3. 3. 3.]]


In [82]:
# what happened to rand_mat when we change sub_mat?
rand_mat

array([[ 3.        ,  3.        ,  3.        , -0.29794316,  0.80229657],
       [ 3.        ,  3.        ,  3.        , -0.42192284, -0.36478021],
       [-0.07835671,  2.11000357, -0.9506834 ,  0.42590556, -0.45060251],
       [-0.16245548, -0.46694343, -0.15550438,  0.31551549, -1.72685098],
       [ 0.70500886, -2.17388943, -0.13136581, -0.73299754,  0.42873511]])

In [85]:
# to prevent this behavior we need to use the .copy() method when we assign sub_mat
# this behavior is the source of MANY errors for early python users!!!

print(rand_mat)

sub_mat = rand_mat[:,:].copy()
sub_mat[:] = 5
print(sub_mat)
print('------------------------------------')
print(rand_mat)

[[ 3.          3.          3.         -0.29794316  0.80229657]
 [ 3.          3.          3.         -0.42192284 -0.36478021]
 [-0.07835671  2.11000357 -0.9506834   0.42590556 -0.45060251]
 [-0.16245548 -0.46694343 -0.15550438  0.31551549 -1.72685098]
 [ 0.70500886 -2.17388943 -0.13136581 -0.73299754  0.42873511]]
[[5. 5. 5. 5. 5.]
 [5. 5. 5. 5. 5.]
 [5. 5. 5. 5. 5.]
 [5. 5. 5. 5. 5.]
 [5. 5. 5. 5. 5.]]
------------------------------------
[[ 3.          3.          3.         -0.29794316  0.80229657]
 [ 3.          3.          3.         -0.42192284 -0.36478021]
 [-0.07835671  2.11000357 -0.9506834   0.42590556 -0.45060251]
 [-0.16245548 -0.46694343 -0.15550438  0.31551549 -1.72685098]
 [ 0.70500886 -2.17388943 -0.13136581 -0.73299754  0.42873511]]


### 2.4 Saving and Loading a NumPy array

**Let's save some NumPy objects on the disk for use later!**

In [86]:
# creating a random matrices
randint_matrix1 = np.random.randint(1,10,10).reshape(2,5)
print(randint_matrix1)
print('')
randint_matrix2 = np.random.randint(10,20,10).reshape(2,5)
print(randint_matrix2)

[[2 3 8 1 7]
 [3 5 3 4 1]]

[[10 12 13 17 12]
 [16 13 16 10 12]]


**Using np.save() function**

In [87]:
np.save('saved_file_name',randint_matrix1)

**Using np.savez() function**

In [88]:
np.savez('multiple_files',randint_matrix1=randint_matrix1,randint_matrix2=randint_matrix2)

* The files will be saved in the directory where the Jupyter Notebook is located.
* With np.save() function, we can save an array/matrix to a NumPy .npy format.
* np.savez() function has an advantage over np.save() function because with np.savez(), we can store several arrays/matrices into a single file in uncompressed .npz format.

In [89]:
# now let's load it
loaded_arr = np.load('saved_file_name.npy')
loaded_multi = np.load('multiple_files.npz')

print(loaded_arr)
print('')
print(loaded_multi)

[[2 3 8 1 7]
 [3 5 3 4 1]]



* We see that .npy file has been loaded but the .npz file is returning a memory location.
* Let's see how to load the values stored in .npz file.

In [90]:
print('1st Matrix: \n',loaded_multi['randint_matrix1'])
print('2nd Matrix: \n',loaded_multi['randint_matrix2'])

new_matrix  = loaded_multi['randint_matrix1']
print('New Matrix: \n',new_matrix)

1st Matrix: 
 [[2 3 8 1 7]
 [3 5 3 4 1]]
2nd Matrix: 
 [[10 12 13 17 12]
 [16 13 16 10 12]]
New Matrix: 
 [[2 3 8 1 7]
 [3 5 3 4 1]]


In [91]:
# we can also save/load text files...but only single variables
np.savetxt('text_file_name.txt',randint_matrix1,delimiter=',')
rand_mat_txt = np.loadtxt('text_file_name.txt',delimiter=',')
print(randint_matrix1)
print('')
print(rand_mat_txt)

[[2 3 8 1 7]
 [3 5 3 4 1]]

[[2. 3. 8. 1. 7.]
 [3. 5. 3. 4. 1.]]
